Client/macos/MacOSBluetoothConnector.mm

```diff
@@ -0,0 +1,161 @@
+		return;
+	}
+	else{
+		fprintf(stderr,"Connected. Yeah!\n");
+	}
+}
+
+-(void)rfcommChannelData:(IOBluetoothRFCOMMChannel *)rfcommChannel data:(void *)dataPointer length:(size_t)dataLength
+{
+	delegateCPP->receivedBytes = (char*)dataPointer;
+	delegateCPP->receivedLength = dataLength;
+	delegateCPP->wantNewData = false;
+}
+
+
+@end
+
+int MacOSBluetoothConnector::send(char* buf, size_t length)
+{
+	int error = [(__bridge IOBluetoothRFCOMMChannel*)rfcommchannel writeSync:(void*)buf length:length];
+	if ( error != kIOReturnSuccess ) {
+		fprintf(stderr,"Error - could send bluetooth command. Error code = %08x.\n",error);
+	}
+	return length;
+}
+
+
+void MacOSBluetoothConnector::connectToMac(MacOSBluetoothConnector* macOSBluetoothConnector)
+{
+	macOSBluetoothConnector->running = 1;
+	//get device
+	IOBluetoothDevice *device = (__bridge IOBluetoothDevice *)macOSBluetoothConnector->rfcommDevice;
+	// create new channel
+	IOBluetoothRFCOMMChannel *channel = [[IOBluetoothRFCOMMChannel alloc] init];
+	// create sppServiceid
+	IOBluetoothSDPUUID *sppServiceUUID = [IOBluetoothSDPUUID uuidWithBytes:(void*)SERVICE_UUID_IN_BYTES length: 16];
+	// get sppServiceRecord
+	IOBluetoothSDPServiceRecord *sppServiceRecord = [device getServiceRecordForUUID:sppServiceUUID];
+	// get rfcommChannelID from sppServiceRecord
+	UInt8 rfcommChannelID;
+	[sppServiceRecord getRFCOMMChannelID:&rfcommChannelID];
+	// setup delegate
+	AsyncCommDelegate* asyncCommDelegate = [[AsyncCommDelegate alloc] init];
+	asyncCommDelegate->delegateCPP = macOSBluetoothConnector;
+	// try to open channel
+	if ( [device openRFCOMMChannelAsync:&channel withChannelID:rfcommChannelID delegate:asyncCommDelegate] != kIOReturnSuccess ) {
+		throw "Error - could not open the rfcomm.\n";
+	}
+	// store the channel
+	macOSBluetoothConnector->rfcommchannel = (__bridge void*) channel;
+
+	printf("Successfully connected");
+
+	// keep thread running
+	while (macOSBluetoothConnector->running) {
+		[[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:1]];
+	}
+}
+void MacOSBluetoothConnector::connect(const std::string& addrStr){
+	// convert mac adress to nsstring
+	NSString *addressNSString = [NSString stringWithCString:addrStr.c_str() encoding:[NSString defaultCStringEncoding]];
+	// get device based on mac adress
+	IOBluetoothDevice *device = [IOBluetoothDevice deviceWithAddressString:addressNSString];
+	// if device is not connected
+	if (![device isConnected]) {
+		[device openConnection];
+	}
+	// store the device in an variable
+	rfcommDevice = (__bridge void*) device;
+	uthread = new std::thread(MacOSBluetoothConnector::connectToMac, this);
+}
+
+int MacOSBluetoothConnector::recv(char* buf, size_t length)
+{
+	// wait for newly received data
+	wantNewData = true;
+
+	// wait until it has received data
+	while (wantNewData) {
+		// run the runLoop so it can actually receive data
+		[runLoop runUntilDate:[NSDate dateWithTimeIntervalSinceNow:10]];
```

Comment:
This has the same problems as the comment I wrote about the `running` flag. Why is the sleep for so long (10 seconds), by the way?
1. Use `std::atomic<bool>`
2. Use a condition variable. 

Same goes for the trickery you're doing here with receivedLength -- it's weird. I think that I would just create an `std::queue` and push message buffers (as `std::vector<unsigned char>`) onto it when they're received. Locking appropriately for inter-thread access, of course. I would also have the predicate of the cond var check whether the queue is empty or not.

Reply:
Got this fixed I guess, could you check?